# K-Nearest Neighbour(KNN) -

In [2]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score 
from sklearn.metrics import confusion_matrix

In [3]:
dataset=pd.read_csv("train.csv")

In [4]:
dataset.head(2)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   

   Parch     Ticket     Fare Cabin Embarked  
0      0  A/5 21171   7.2500   NaN        S  
1      0   PC 17599  71.2833   C85        C

In [5]:
dataset.columns 

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [6]:
dataset.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         0
dtype: int64

In [7]:
dataset1=dataset.drop(["Name","Ticket","Cabin"],axis=1)

In [8]:
dataset1.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch',
       'Fare', 'Embarked'],
      dtype='object')

In [9]:
le=LabelEncoder()
le.fit(dataset1["Sex"])

LabelEncoder()

In [10]:
dataset1["Sex"]=le.transform(dataset1["Sex"])

In [11]:
print(le.classes_)

['female' 'male']


In [12]:
le.fit(dataset1["Embarked"])

LabelEncoder()

In [13]:
print(le.classes_)

['C' 'Q' 'S']


In [14]:
dataset1["Embarked"]=le.transform(dataset1["Embarked"])

In [15]:
print(le.classes_)

['C' 'Q' 'S']


In [16]:
from sklearn import neighbors

In [17]:
KNN_dataset=dataset1

In [18]:
KNN_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 889 entries, 0 to 888
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  889 non-null    int64  
 1   Survived     889 non-null    int64  
 2   Pclass       889 non-null    int64  
 3   Sex          889 non-null    int32  
 4   Age          889 non-null    float64
 5   SibSp        889 non-null    int64  
 6   Parch        889 non-null    int64  
 7   Fare         889 non-null    float64
 8   Embarked     889 non-null    int32  
dtypes: float64(2), int32(2), int64(5)
memory usage: 55.7 KB


In [19]:
KNN_dataset.drop_duplicates()

PassengerId  Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked
0              1         0       3    1  22.0      1      0   7.2500         2
1              2         1       1    0  38.0      1      0  71.2833         0
2              3         1       3    0  26.0      0      0   7.9250         2
3              4         1       1    0  35.0      1      0  53.1000         2
4              5         0       3    1  35.0      0      0   8.0500         2
..           ...       ...     ...  ...   ...    ...    ...      ...       ...
884          887         0       2    1  27.0      0      0  13.0000         2
885          888         1       1    0  19.0      0      0  30.0000         2
886          889         0       3    0  45.0      1      2  23.4500         2
887          890         1       1    1  26.0      0      0  30.0000         0
888          891         0       3    1  32.0      0      0   7.7500         1

[889 rows x 9 columns]

In [20]:
KNN_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 889 entries, 0 to 888
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  889 non-null    int64  
 1   Survived     889 non-null    int64  
 2   Pclass       889 non-null    int64  
 3   Sex          889 non-null    int32  
 4   Age          889 non-null    float64
 5   SibSp        889 non-null    int64  
 6   Parch        889 non-null    int64  
 7   Fare         889 non-null    float64
 8   Embarked     889 non-null    int32  
dtypes: float64(2), int32(2), int64(5)
memory usage: 55.7 KB


In [21]:
y=KNN_dataset["Pclass"]
X=KNN_dataset.drop(["Pclass","PassengerId"],axis=1)

In [22]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.3,random_state=0)

In [23]:
knn=neighbors.KNeighborsClassifier(n_neighbors=3)

In [24]:
knn.fit(X_train,y_train).score(X_test,y_test)*100

86.51685393258427

In [25]:
y_pred=knn.predict(X_test)

In [26]:
confusion_matrix(y_test,y_pred)

array([[ 63,   5,   2],
       [  7,  28,  14],
       [  3,   5, 140]], dtype=int64)

In [75]:
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
def K_near():
    for i in range(1,268):
        knn=neighbors.KNeighborsClassifier(n_neighbors=i)
        acc_score=knn.fit(X_train,y_train).score(X_test,y_test)*100
        Acc=np.array(acc_score)
        print("The accuracy of the model we developed with respect to k=", i, "value is:\n",Acc,"%")
        y_pred=knn.predict(X_test)
        print("The respective confusion matrix corresponding to K-value is:\n", confusion_matrix(y_test,y_pred))
        i=i+1
K_near()

The accuracy of the model we developed with respect to k= 1 value is:
 89.13857677902621 %
The respective confusion matrix corresponding to K-value is:
 [[ 62   6   2]
 [  3  37   9]
 [  1   8 139]]
The accuracy of the model we developed with respect to k= 2 value is:
 86.51685393258427 %
The respective confusion matrix corresponding to K-value is:
 [[ 67   2   1]
 [  8  37   4]
 [  4  17 127]]
The accuracy of the model we developed with respect to k= 3 value is:
 86.51685393258427 %
The respective confusion matrix corresponding to K-value is:
 [[ 63   5   2]
 [  7  28  14]
 [  3   5 140]]
The accuracy of the model we developed with respect to k= 4 value is:
 82.39700374531836 %
The respective confusion matrix corresponding to K-value is:
 [[ 61   7   2]
 [  8  30  11]
 [  4  15 129]]
The accuracy of the model we developed with respect to k= 5 value is:
 83.14606741573034 %
The respective confusion matrix corresponding to K-value is:
 [[ 58  10   2]
 [  7  28  14]
 [  3   9 136]]
The a

The respective confusion matrix corresponding to K-value is:
 [[ 61   7   2]
 [  7  19  23]
 [  7  12 129]]
The accuracy of the model we developed with respect to k= 46 value is:
 78.65168539325843 %
The respective confusion matrix corresponding to K-value is:
 [[ 61   7   2]
 [  7  20  22]
 [  7  12 129]]
The accuracy of the model we developed with respect to k= 47 value is:
 77.52808988764045 %
The respective confusion matrix corresponding to K-value is:
 [[ 61   6   3]
 [  7  17  25]
 [  7  12 129]]
The accuracy of the model we developed with respect to k= 48 value is:
 77.52808988764045 %
The respective confusion matrix corresponding to K-value is:
 [[ 61   6   3]
 [  7  17  25]
 [  7  12 129]]
The accuracy of the model we developed with respect to k= 49 value is:
 77.52808988764045 %
The respective confusion matrix corresponding to K-value is:
 [[ 62   6   2]
 [  7  16  26]
 [  7  12 129]]
The accuracy of the model we developed with respect to k= 50 value is:
 77.15355805243446 %


 [[ 54   6  10]
 [  5  10  34]
 [  5   7 136]]
The accuracy of the model we developed with respect to k= 90 value is:
 75.65543071161048 %
The respective confusion matrix corresponding to K-value is:
 [[ 54   5  11]
 [  5  10  34]
 [  5   5 138]]
The accuracy of the model we developed with respect to k= 91 value is:
 75.65543071161048 %
The respective confusion matrix corresponding to K-value is:
 [[ 54   6  10]
 [  5  10  34]
 [  5   5 138]]
The accuracy of the model we developed with respect to k= 92 value is:
 75.65543071161048 %
The respective confusion matrix corresponding to K-value is:
 [[ 54   6  10]
 [  5  10  34]
 [  5   5 138]]
The accuracy of the model we developed with respect to k= 93 value is:
 75.28089887640449 %
The respective confusion matrix corresponding to K-value is:
 [[ 54   6  10]
 [  5  10  34]
 [  5   6 137]]
The accuracy of the model we developed with respect to k= 94 value is:
 75.28089887640449 %
The respective confusion matrix corresponding to K-value is:


The accuracy of the model we developed with respect to k= 132 value is:
 76.02996254681648 %
The respective confusion matrix corresponding to K-value is:
 [[ 53  13   4]
 [  5  11  33]
 [  5   4 139]]
The accuracy of the model we developed with respect to k= 133 value is:
 76.02996254681648 %
The respective confusion matrix corresponding to K-value is:
 [[ 53  13   4]
 [  5  11  33]
 [  5   4 139]]
The accuracy of the model we developed with respect to k= 134 value is:
 75.65543071161048 %
The respective confusion matrix corresponding to K-value is:
 [[ 52  13   5]
 [  5  11  33]
 [  5   4 139]]
The accuracy of the model we developed with respect to k= 135 value is:
 76.02996254681648 %
The respective confusion matrix corresponding to K-value is:
 [[ 52  13   5]
 [  5  11  33]
 [  5   3 140]]
The accuracy of the model we developed with respect to k= 136 value is:
 76.02996254681648 %
The respective confusion matrix corresponding to K-value is:
 [[ 52  13   5]
 [  5  11  33]
 [  5   3 1

 74.90636704119851 %
The respective confusion matrix corresponding to K-value is:
 [[ 50   8  12]
 [  5   7  37]
 [  5   0 143]]
The accuracy of the model we developed with respect to k= 175 value is:
 75.65543071161048 %
The respective confusion matrix corresponding to K-value is:
 [[ 52   6  12]
 [  5   7  37]
 [  5   0 143]]
The accuracy of the model we developed with respect to k= 176 value is:
 75.65543071161048 %
The respective confusion matrix corresponding to K-value is:
 [[ 52   6  12]
 [  5   7  37]
 [  5   0 143]]
The accuracy of the model we developed with respect to k= 177 value is:
 75.65543071161048 %
The respective confusion matrix corresponding to K-value is:
 [[ 52   6  12]
 [  5   7  37]
 [  5   0 143]]
The accuracy of the model we developed with respect to k= 178 value is:
 75.65543071161048 %
The respective confusion matrix corresponding to K-value is:
 [[ 52   6  12]
 [  5   7  37]
 [  5   0 143]]
The accuracy of the model we developed with respect to k= 179 value

The respective confusion matrix corresponding to K-value is:
 [[ 48   4  18]
 [  4   2  43]
 [  5   0 143]]
The accuracy of the model we developed with respect to k= 216 value is:
 72.28464419475655 %
The respective confusion matrix corresponding to K-value is:
 [[ 48   4  18]
 [  4   2  43]
 [  5   0 143]]
The accuracy of the model we developed with respect to k= 217 value is:
 72.28464419475655 %
The respective confusion matrix corresponding to K-value is:
 [[ 48   4  18]
 [  4   2  43]
 [  5   0 143]]
The accuracy of the model we developed with respect to k= 218 value is:
 72.28464419475655 %
The respective confusion matrix corresponding to K-value is:
 [[ 48   4  18]
 [  4   2  43]
 [  5   0 143]]
The accuracy of the model we developed with respect to k= 219 value is:
 72.28464419475655 %
The respective confusion matrix corresponding to K-value is:
 [[ 48   2  20]
 [  4   2  43]
 [  5   0 143]]
The accuracy of the model we developed with respect to k= 220 value is:
 72.284644194756

# Conclusion - 

**Considering all the K-values possible from 1 to 267 we have found that the model has the maximum value when K=1 having the Accuracy Score of 89.13857677902621 %**

# Support Vector Machine(SVM) - 

In [78]:
KNN_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 889 entries, 0 to 888
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  889 non-null    int64  
 1   Survived     889 non-null    int64  
 2   Pclass       889 non-null    int64  
 3   Sex          889 non-null    int32  
 4   Age          889 non-null    float64
 5   SibSp        889 non-null    int64  
 6   Parch        889 non-null    int64  
 7   Fare         889 non-null    float64
 8   Embarked     889 non-null    int32  
dtypes: float64(2), int32(2), int64(5)
memory usage: 55.7 KB


In [79]:
SVM_dataset=KNN_dataset

In [80]:
SVM_dataset.head(2)

PassengerId  Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked
0            1         0       3    1  22.0      1      0   7.2500         2
1            2         1       1    0  38.0      1      0  71.2833         0

In [87]:
SVM_dataset.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch',
       'Fare', 'Embarked'],
      dtype='object')

In [95]:
Attributes = ['Pclass', 'Survived', 'Sex', 'SibSp', 'Parch', 'Embarked']

In [96]:
from sklearn import svm
def SVM(x,y):
    x_train,x_test,y_train,y_test = train_test_split(x, y, test_size = 0.3, random_state = 0)
    clf = svm.SVC(gamma = 0.01)
    clf.fit(x_train,y_train)
    y_pred = clf.predict(x_test)
    print("Accuracy_score:",accuracy_score(y_test, y_pred))
    print("Confusion-matrix:")
    print(confusion_matrix(y_test, y_pred))

In [97]:
for DV in Attributes:
    y = SVM_dataset[DV]
    x = SVM_dataset.drop(DV,axis=1)
    print("")
    print(DV,"Vs Remaining Dv")
    SVM(x,y)


Pclass Vs Remaining Dv
Accuracy_score: 0.5617977528089888
Confusion-matrix:
[[  9   1  60]
 [  2   4  43]
 [  3   8 137]]

Survived Vs Remaining Dv
Accuracy_score: 0.550561797752809
Confusion-matrix:
[[136  21]
 [ 99  11]]

Sex Vs Remaining Dv
Accuracy_score: 0.5880149812734082
Confusion-matrix:
[[ 10  88]
 [ 22 147]]

SibSp Vs Remaining Dv
Accuracy_score: 0.6779026217228464
Confusion-matrix:
[[178   4   0   0   0   0   0]
 [ 62   3   0   0   0   0   0]
 [  7   1   0   0   0   0   0]
 [  6   1   0   0   0   0   0]
 [  1   1   0   0   0   0   0]
 [  1   0   0   0   0   0   0]
 [  2   0   0   0   0   0   0]]

Parch Vs Remaining Dv
Accuracy_score: 0.7415730337078652
Confusion-matrix:
[[197   0   1   0   0]
 [ 39   1   0   0   0]
 [ 27   0   0   0   0]
 [  1   0   0   0   0]
 [  1   0   0   0   0]]

Embarked Vs Remaining Dv
Accuracy_score: 0.7116104868913857
Confusion-matrix:
[[  0   2  47]
 [  2   1  19]
 [  6   1 189]]
